src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,951 @@
+            bool overload, CancellationToken cancellationToken)
+        {
+            var modifiedSolution = originalDocument.Project.Solution;
+            var mappingDictionary = await MapExpressionToParametersAsync(originalDocument, expression, cancellationToken).ConfigureAwait(false);
+
+            foreach (var (project, projectCallSites) in callSites.GroupBy(kvp => kvp.Key.Project))
+            {
+                var compilation = await project.GetRequiredCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var (document, invocations) in projectCallSites)
+                {
+                    if (trampoline || overload)
+                    {
+                        var newRoot = await ModifyDocumentInvocationsTrampolineOverloadAndIntroduceParameterAsync(compilation,
+                            document, originalDocument, invocations, mappingDictionary, methodSymbol, containingMethod,
+                            allOccurrences, parameterName, expression, trampoline, overload, cancellationToken).ConfigureAwait(false);
+                        modifiedSolution = modifiedSolution.WithDocumentSyntaxRoot(originalDocument.Id, newRoot);
+                    }
+                    else
+                    {
+                        var newRoot = await ModifyDocumentInvocationsAndIntroduceParameterAsync(compilation,
+                            originalDocument, document, mappingDictionary, methodSymbol, containingMethod,
+                            expression, allOccurrences, parameterName, invocations,
+                            cancellationToken).ConfigureAwait(false);
+                        modifiedSolution = modifiedSolution.WithDocumentSyntaxRoot(originalDocument.Id, newRoot);
+                    }
+                }
+            }
+
+            return modifiedSolution;
+        }
+
+        /// <summary>
+        /// For the trampoline case, it goes through the invocations and adds an argument which is a 
+        /// call to the extracted method.
+        /// Introduces a new method overload or new trampoline method.
+        /// Updates the original method site with a newly introduced parameter.
+        /// 
+        /// ****Trampoline Example:****
+        /// public void M(int x, int y)
+        /// {
+        ///     int f = [|x * y|];
+        ///     Console.WriteLine(f);
+        /// }
+        /// 
+        /// public void InvokeMethod()
+        /// {
+        ///     M(5, 6);
+        /// }
+        /// 
+        /// ---------------------------------------------------->
+        /// 
+        /// public int GetF(int x, int y) // Generated method
+        /// {
+        ///     return x * y;
+        /// }
+        /// 
+        /// public void M(int x, int y, int f)
+        /// {
+        ///     Console.WriteLine(f);
+        /// }
+        /// 
+        /// public void InvokeMethod()
+        /// {
+        ///     M(5, 6, GetF(5, 6)); //Fills in with call to generated method
```

Comment:
do you have tests where you start with:

`x.y.M(5, 6)`?  In that test,  we should end up with `x.y.M(5, 6, x.y.GetF(5, 6))`.

IF no tests, please add to validate this scenario.

Reply:
I do not, will ask about this scenario.

Reply:
has this been dealt with?

Reply:
Yes